### DISCUSSION TITLE
Kirim WhatsApp dengan API KirimWA.id dan PHP - Error

### DISCUSSION THREAD

**Discussion by mohamadlukman** [kirimwa-id/forum#128] [Tanya Jawab] [answered]
Selamat sing rekan, saya orang awam neh di progeming 
saya  mau tanya kenapa pas di jalanin sourcecode nya masih ajah error, padahal udh masukan code sesuai dengan tahapan2, namun pas di di test masih ada ada error

Referensi Link:
https://teknocerdas.com/programming/kirim-whatsapp-dengan-api-kirimwa-id-dan-php/

![image](https://user-images.githubusercontent.com/110963624/183828664-e02b4c80-4ef1-4168-a983-108f899bcdbe.png)

Mohon  pencerahanya dan info nya

**Comment by melodiazz** [comment]
Boleh dipost disini untuk syntaxnya pada add-device.php dan blur untuk informasi sensitive seperti token dll ya.

**Reply by mohamadlukman** [reply]
<?php

/**
 * Make a request to API KirimWA.id
 *
 * @param Array $params
 * @return Array
 * @throws Exception
 */
function apiKirimWaRequest(array $params) {
  $httpStreamOptions = [
    'method' => $params['method'] ?? 'GET',
    'header' => [
      'Content-Type: application/json',
      'Authorization: Bearer ' . ($params['token'] ?? '')
    ],
    'timeout' => 15,
    'ignore_errors' => true
  ];

  if ($httpStreamOptions['method'] === 'POST') {
    $httpStreamOptions['header'][] = sprintf('Content-Length: %d', strlen($params['payload'] ?? ''));
    $httpStreamOptions['content'] = $params['payload'];
  }

  // Join the headers using CRLF
  $httpStreamOptions['header'] = implode("\r\n", $httpStreamOptions['header']) . "\r\n";

  $stream = stream_context_create(['http' => $httpStreamOptions]);
  $response = file_get_contents($params['url'], false, $stream);

  // Headers response are created magically and injected into
  // variable named $http_response_header
  $httpStatus = $http_response_header[0];

  preg_match('#HTTP/[\d\.]+\s(\d{3})#i', $httpStatus, $matches);

  if (! isset($matches[1])) {
    throw new Exception('Can not fetch HTTP response header.');
  }

  $statusCode = (int)$matches[1];
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
  }

  throw new Exception($response, $statusCode);
}

// Lihat apiKirimWaRequest() pada Contoh Integrasi diatas
try {
  $reqParams = [
    'token' => 'TOKEN_ANDA',
    'url' => 'https://api.kirimwa.id/v1/devices',
    'method' => 'POST',
    'payload' => json_encode([
      'device_id' => 'iphone-x-pro' // ganti sesuai keinginan
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

**Reply by melodiazz** [reply]
sepertinya ini copas full dari source ya, seharusnya banyak yang disesuaikan seperti 'TOKEN_ANDA', 'iphone-x-pro', dll.

Saran saya coba pahami dulu programming khususnya php sebelum mulai menggunakan fitur kirimwa ini.

salah satu video yang bisa memantu belajar php : https://www.youtube.com/watch?v=TaBWhb5SPfc

**Reply by mohamadlukman** [reply]
sudah di sesuikan pake token yg saya dapat,
namun tetep ga bisa..

**Comment by mohamadlukman** [comment]
**add-device.php**

<?php

/**
 * Make a request to API KirimWA.id
 *
 * @param Array $params
 * @return Array
 * @throws Exception
 */
function apiKirimWaRequest(array $params) {
  $httpStreamOptions = [
    'method' => $params['method'] ?? 'GET',
    'header' => [
      'Content-Type: application/json',
      'Authorization: Bearer ' . ($params['token'] ?? '')
    ],
    'timeout' => 15,
    'ignore_errors' => true
  ];

  if ($httpStreamOptions['method'] === 'POST') {
    $httpStreamOptions['header'][] = sprintf('Content-Length: %d', strlen($params['payload'] ?? ''));
    $httpStreamOptions['content'] = $params['payload'];
  }

  // Join the headers using CRLF
  $httpStreamOptions['header'] = implode("\r\n", $httpStreamOptions['header']) . "\r\n";

  $stream = stream_context_create(['http' => $httpStreamOptions]);
  $response = file_get_contents($params['url'], false, $stream);

  // Headers response are created magically and injected into
  // variable named $http_response_header
  $httpStatus = $http_response_header[0];

  preg_match('#HTTP/[\d\.]+\s(\d{3})#i', $httpStatus, $matches);

  if (! isset($matches[1])) {
    throw new Exception('Can not fetch HTTP response header.');
  }

  $statusCode = (int)$matches[1];
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
  }

  throw new Exception($response, $statusCode);
}

// Lihat apiKirimWaRequest() pada Contoh Integrasi diatas
try {
  $reqParams = [
    'token' => 'TOKEN_ANDA',
    'url' => 'https://api.kirimwa.id/v1/devices',
    'method' => 'POST',
    'payload' => json_encode([
      'device_id' => 'iphone-x-pro' // ganti sesuai keinginan
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

**Reply by kafaa-dev** [reply]
Pakai versi berapa ya mas

**Reply by mohamadlukman** [reply]
pake Versi PHP: 5.6.40

**Reply by kafaa-dev** [reply]
oh pantesan

**Reply by mohamadlukman** [reply]
Harusnya versi brp?

**Reply by kafaa-dev** [reply]
sebenarnya versi segitu juga bisa, hanya ada beberapa kode yang harus diubah

**Reply by kafaa-dev** [reply]
sudah bisa mas?

**Comment by kafaa-dev** [comment]
Kode di tutorial [TeknoCerdas](https://teknocerdas.com/programming/kirim-whatsapp-dengan-api-kirimwa-id-dan-php/) menggunakan code yang minimal versi php 7.

Alasannya adalah karena fitur "null coalescing operator" baru ditambahkan di php versi 7

Silahkan ganti kode fungsi `apiKirimWaRequest` dengan :
```php
/**
 * Make a request to API KirimWA.id
 *
 * @param Array $params
 * @return Array
 * @throws Exception
 */
function apiKirimWaRequest(array $params) {
  $httpStreamOptions = [
    'method' => isset($params['method']) ? $params['method'] : 'GET',
    'header' => [
      'Content-Type: application/json',
      'Authorization: Bearer ' . (isset($params['token']) ? $params['token'] : '')
    ],
    'timeout' => 15,
    'ignore_errors' => true
  ];

  if ($httpStreamOptions['method'] === 'POST') {
    $httpStreamOptions['header'][] = sprintf('Content-Length: %d', strlen(isset($params['payload']) ? $params['payload'] : ''));
    $httpStreamOptions['content'] = $params['payload'];
  }

  // Join the headers using CRLF
  $httpStreamOptions['header'] = implode("\r\n", $httpStreamOptions['header']) . "\r\n";

  $stream = stream_context_create(['http' => $httpStreamOptions]);
  $response = file_get_contents($params['url'], false, $stream);

  // Headers response are created magically and injected into
  // variable named $http_response_header
  $httpStatus = $http_response_header[0];

  preg_match('#HTTP/[\d\.]+\s(\d{3})#i', $httpStatus, $matches);

  if (! isset($matches[1])) {
    throw new Exception('Can not fetch HTTP response header.');
  }

  $statusCode = (int)$matches[1];
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
  }

  throw new Exception($response, $statusCode);
}
```

Semoga membantu 🙏

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/183835454-faef55b0-f6bc-4eee-8d66-8b0404dd6230.png)


jadi ngeblank

**Reply by kafaa-dev** [reply]
mohon sertakan kode yang lengkap, apa yang diinginkan? apa tujuan yang ingin dicapai?

> Penting! Jangan mencantumkan informasi yang bersifat rahasia seperti API Token, password, nomor telepon dan sebagainya.

**Reply by mohamadlukman** [reply]
Kodenya sama kaya mas posting di atas, 
tujuanya pengen bisa kirim notif pakai API Wa.

**Reply by mohamadlukman** [reply]
nah ternyata ada tutornya di [TeknoCerdas](https://teknocerdas.com/programming/kirim-whatsapp-dengan-api-kirimwa-id-dan-php/) namun pas di ikutin ga jalan, apa mungkin ada kesalah.

**Reply by melodiazz** [reply]
mas @mohamadlukman ini kami mau coba bantu tapi kalau masnya gak bisa copy paste code yang mas gunakan di file phpnya (bukan copas kode teknocerdas), maka kita juga sulit mau bantu.

kita perlu cek kode yang mas pakai untuk bisa tahu letak kesalahannya ada dimana.

**Reply by mohamadlukman** [reply]
functions.php

`<?php

/**
 * Make a request to API KirimWA.id
 *
 * @param Array $params
 * @return Array
 * @throws Exception
 */
function apiKirimWaRequest(array $params) {
   $httpStreamOptions = [
    'method' => isset($params['method']) ? $params['method'] : 'GET',
    'header' => [
      'Content-Type: application/json',
      'Authorization: Bearer ' . (isset($params['token']) ? $params['token'] : '')
    ],
    'timeout' => 15,
    'ignore_errors' => true
  ];

  if ($httpStreamOptions['method'] === 'POST') {
    $httpStreamOptions['header'][] = sprintf('Content-Length: %d', strlen(isset($params['payload']) ? $params['payload'] : ''));
    $httpStreamOptions['content'] = $params['payload'];
  }

  // Join the headers using CRLF
  $httpStreamOptions['header'] = implode("\r\n", $httpStreamOptions['header']) . "\r\n";

  $stream = stream_context_create(['http' => $httpStreamOptions]);
  $response = file_get_contents($params['url'], false, $stream);

  // Headers response are created magically and injected into
  // variable named $http_response_header
  $httpStatus = $http_response_header[0];

  preg_match('#HTTP/[\d\.]+\s(\d{3})#i', $httpStatus, $matches);

  if (! isset($matches[1])) {
    throw new Exception('Can not fetch HTTP response header.');
  }

  $statusCode = (int)$matches[1];
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
  }

  throw new Exception($response, $statusCode);
}`
===========================================================
add-device.php

`<?php
require './functions.php';

try {
  $baseApiUrl = getenv('API_BASEURL') ? getenv('API_BASEURL') : 'https://api.kirimwa.id/v1';
  $reqParams = [
    'token' => getenv('fZTnDy1l1HH4dAMgyxxxxxxxxxxxx'),
    'url' => $baseApiUrl . '/devices',
    'method' => 'POST',
    'payload' => json_encode([
      'device_id' => getenv('DEVICE_ID')
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}`

============================

mas @melodiazz kaya gini mas .... 🙏🙏

**Reply by melodiazz** [reply]
oke kalau memang kode yang dipakai **persis** seperti yang dipost. Maka yang perlu dilakukan adalah:

functions.php :
1. Tambahkan ?> pada akhir file

sehingga setelah diubah hasilnya sebagai berikut:

![image](https://user-images.githubusercontent.com/22655268/183839660-4ef718e8-f935-4aa6-9acf-c212c4eb5d69.png)

add-device.php :

1. require './functions.php'; -> diubah menjadi require 'functions.php'; 
(dengan asumsi functions.php & add-device.php ada dalam directory yang sama ya)
2. 'token' => getenv('fZTnDy1l1HH4dAMgyxxxxxxxxxxxx'), diganti menjadi 'token' => 'fZTnDy1l1HH4dAMgyxxxxxxxxxxxx',
karna bila menggunakan getenv maka seharusnya getenv['API_TOKEN'], sementara fZTnDy1l1HH4dAMgyxxxxxxxxxxxx adalah token itu sendiri sehingga tidak perlu menggunakan getenv
3. 'device_id' => getenv('DEVICE_ID') diganti menjadi 'device_id' => 'namadevice'
namadevice silahkan disesuaikan dengan keinginan.
4. Tambahkan ?> pada akhir file

sehingga setelah diubah hasilnya sebagai berikut:
![image](https://user-images.githubusercontent.com/22655268/183841690-b389fac9-07f9-4ce1-ae1a-dd31b3129dfd.png)

**Reply by kafaa-dev** [reply]
@melodiazz sebenarnya `?>` pada php itu tidak wajib kecuali ingin memisahkan dengan kode html

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/183845673-a05602b6-7bad-4e80-9998-95d9bdd494d3.png)

![image](https://user-images.githubusercontent.com/110963624/183845870-ed640fba-d534-4c1e-897e-f3604cf00fb7.png)

**Reply by melodiazz** [reply]
coba line 8 diganti jadi seperti ini

``` 'url' => 'https://api.kirimwa.id/v1/devices',```

**Reply by mohamadlukman** [reply]
tetep mas errornya di line 16
sudh di ganti dengan '**url' => 'https://api.kirimwa.id/v1/devices',**

**Reply by melodiazz** [reply]
functions.php

sebelum ?> ditambah karakter } mas

jadi seperti ini
![image](https://user-images.githubusercontent.com/22655268/183857644-8b44214d-d76b-4519-8d29-a6d5f1eb4a45.png)

**Comment by mohamadlukman** [comment]
![image](https://user-images.githubusercontent.com/110963624/183860093-a51cf226-9f41-4abe-85d9-4b8e185dafb8.png)

Sudah di tambahkan *}* masih tetep ada erorr di line 16

**Reply by melodiazz** [reply]
line 16 pada add-device.php coba diubah jadi seperti ini mas

```$respon = isset($response['body']) ? $response['body'] : ''; echo $respon;```

**Reply by kafaa-dev** [reply]
pada file `functions.php` tambahkan baris :
```php
throw new Exception($response, $statusCode);
```

setelah 
```php
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
}
```

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/184074813-c4e03b1c-b033-4ca2-86b3-29779aefef23.png)


============================================================
**functions.php**

`<?php

/**
 * Make a request to API KirimWA.id
 *
 * @param Array $params
 * @return Array
 * @throws Exception
 */
function apiKirimWaRequest(array $params) {
   $httpStreamOptions = [
    'method' => isset($params['method']) ? $params['method'] : 'GET',
    'header' => [
      'Content-Type: application/json',
      'Authorization: Bearer ' . (isset($params['token']) ? $params['token'] : '')
    ],
    'timeout' => 15,
    'ignore_errors' => true
  ];

  if ($httpStreamOptions['method'] === 'POST') {
    $httpStreamOptions['header'][] = sprintf('Content-Length: %d', strlen(isset($params['payload']) ? $params['payload'] : ''));
    $httpStreamOptions['content'] = $params['payload'];
  }

  // Join the headers using CRLF
  $httpStreamOptions['header'] = implode("\r\n", $httpStreamOptions['header']) . "\r\n";

  $stream = stream_context_create(['http' => $httpStreamOptions]);
  $response = file_get_contents($params['url'], false, $stream);

  // Headers response are created magically and injected into
  // variable named $http_response_header
  $httpStatus = $http_response_header[0];

  preg_match('#HTTP/[\d\.]+\s(\d{3})#i', $httpStatus, $matches);

  if (! isset($matches[1])) {
    throw new Exception('Can not fetch HTTP response header.');
  }

  $statusCode = (int)$matches[1];
  if ($statusCode >= 200 && $statusCode < 300) {
    return ['body' => $response, 'statusCode' => $statusCode, 'headers' => $http_response_header];
}

throw new Exception($response, $statusCode);

}

?>
`
===============================================

**add-device.php**

`<?php
require 'functions.php';

try {
  $baseApiUrl = getenv('API_BASEURL') ? getenv('API_BASEURL') : 'https://api.kirimwa.id/v1';
  $reqParams = [
    'token' => 'fZTnDy1l1HH4dAMgy8EQ79iRgEzB9XBOGZEgRv8gKgpp95pI9FXNvzL43p5BYbec-mohamad',
    'url' => 'https://api.kirimwa.id/v1/devices',
    'method' => 'POST',
    'payload' => json_encode([
    'device_id' => 'Asus_Hp'
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  $respon = isset($response['body']) ? $response['body'] : ''; echo $respon;
} catch (Exception $e) {
  print_r($e);
}
?>`

Masih error mas se..

**Reply by melodiazz** [reply]
Sesuai pesan error, maka Asus_HP perlu diganti menjadi asus-hp

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/184075556-9fbb6fb7-50ab-49c9-a60f-6fff44fff8a6.png)

**Reply by melodiazz** [reply]
nice, itu sudah berhasil add device mas

**Reply by mohamadlukman** [reply]
iya mas...
knp statusnya Disconnected ya?

**Reply by melodiazz** [reply]
iya karna kan baru tahap add device, masih ada tahap selanjutnya sesuai tutorial yang tersedia di teknocerdas. Dilanjutkan saja mas ikuti tutorialnya.

**Comment by rioastamal** [comment]
@mohamadlukman Mohon apit kode dengan triple backtick biar mudah dibaca orang lain.

````
```
// kode didalam sini
```
````

@melodiazz @syahid246 Terima kasih sudah membantu jawab.

**Comment by mohamadlukman** [comment]
![image](https://user-images.githubusercontent.com/110963624/184081859-a52315fb-d865-4ee6-bc80-3612b9a0211c.png)

=========================================================
**scan-qr.php**
`<?php
require 'functions.php';

try {
  $query = http_build_query(['device_id' => getenv('DEVICE_ID')]);
  $baseApiUrl = getenv('API_BASEURL') ? getenv('API_BASEURL') : 'https://api.kirimwa.id/v1';
  $reqParams = [
    'token' => 'fZTnDy1l1HH4dAxxxxxxx',
    'url' => 'https://api.kirimwa.id/v1/qr',
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}`

**Reply by mohamadlukman** [reply]
Sudah di coba di ganti bagin **device_id** menjadi *asus-hp* masih muncul erorr seperti di atas
 $query = http_build_query(['device_id' => getenv('DEVICE_ID')]);

**Reply by melodiazz** [reply]
```$query = http_build_query(['device_id' => 'asus-hp']);```

**Reply by melodiazz** [reply]
bagian ini ikuti sama persis dengan tutorial dari teknocerdas saja gpp mas
```'url' => sprintf('%s/qr?%s', $baseApiUrl, $query)```

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/184083055-d78cb05f-df8e-4791-8b75-2d0ddc148c13.png)

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/184083343-589ae8f4-108c-4e8c-b730-df4b59da9b4c.png)

Siipp muncull

**Comment by mohamadlukman** [comment]
![image](https://user-images.githubusercontent.com/110963624/184274729-eeba54a0-4602-4dcc-a4c7-9fab3e48773d.png)

Saya coba test massage tapi notif nya pending, itu salah dari coding saya atau gmn ya?

**send-message.php**

`<?php
require 'functions.php';

try {
  $baseApiUrl = getenv('API_BASEURL') ? getenv('API_BASEURL') : 'https://api.kirimwa.id/v1';
  $reqParams = [
    'token' => 'fZTnDy1l1HH4xxxxxxxxxxxxxxxxxxxxx',
    'url' => $baseApiUrl . '/messages',
    'method' => 'POST',
    'payload' => json_encode([
      'message' => 'Halo ini adalah pesan dari api.kirimwa.id',
      'phone_number' => '6285781332550',
      'message_type' => 'text',
      'device_id' => 'asus-hp'
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}`

**Reply by melodiazz** [reply]
gak salah mas, memang awalnya pasti pending. tapi nanti harusnya masuk kok messagenya

**Reply by mohamadlukman** [reply]
cara kasih no tlp formatnya udh bener ya
ga harus pake +62?

**Reply by melodiazz** [reply]
iya sudah benar seperti itu formatnya

**Reply by mohamadlukman** [reply]
sudah ulang2 kirim pesan tapi ga terkirim jg ya?

**Reply by mohamadlukman** [reply]
![image](https://user-images.githubusercontent.com/110963624/184331113-c62b0d3a-0ef3-4398-97eb-c3cf579cf56e.png)

Alhamdulilah sudah bisa.... kirm

**Comment by mohamadlukman** [comment, marked as answer]
https://drive.google.com/drive/folders/1rkAuzpkq7_u2bsp2MNB3r4yF0ExXxGso?usp=sharing

barangkali ada yg mau....
tinggal masukan API Wa masing2

**Reply by DNDshippuden** [reply]
Terima kasih mas... sangat membantu.
